Cytoscape Popper Suggestions

hmtinc opened the pull request:
- #2 - Deleted remaining DS_Store 
- #3 - Moved .js files out of core and created a collection.js file
- #4 - Used Error instead of Exception
- #5 - Renamed Parameters in getHTMLObject
- #6 - Renamed dim to bounding box 
- #7 - Renamed all instances of cyElement to target 
- #8 - Generalized the functions and made isNode a function instead of a boolean
- #9 - Fix Indentation

maxkfranz commented on src/createBoundingBox.js:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    //Override defaults with 3 if popper is called on core
+    if(userOptions.core){
```
You should instead be setting a boundingBox for the core case.  No function but the top-level `cy.popperRef()` or `ele.popperRef()` should have node/edge/core-specific code.

hmtinc replied:
Removed

maxkfranz commented on src/createReferenceObject.js:
```diff
@@ -1,19 +1,25 @@
 
 //Create a reference object for an element
-module.exports.getRef= function (cyElement, userOptions) {
+module.exports.getRef = function (target, userOptions) {
```
This function should not have if-else code for nodes/edges/core.  It should have everything passed into it.  Only the topmost functions should have core/node/edge-specific code.

hmtinc replied:
Removed

hmtinc commented:
@maxkfranz  I have removed all the node/edge/core specific code from the low level functions. I moved any required checks into the higher-level functions in core.js and collection.js

maxkfranz commented:
This function is not necessary:

https://github.com/hmtinc/cytoscape.js-popper/blob/b2be9e1c1b617885428058fbdb83a9c26d99646c/src/common.js#L19

You should be using `assign()` when you create your internal options object.  It should never be null.

maxkfranz commented:
This function should not exist:

https://github.com/hmtinc/cytoscape.js-popper/blob/b2be9e1c1b617885428058fbdb83a9c26d99646c/src/common.js#L2-L16

That sort of code should exist *only* in `cy.popperRef()` and `ele.popperRef()`.

maxkfranz commented:
This shouldn't exist: 

https://github.com/hmtinc/cytoscape.js-popper/blob/b2be9e1c1b617885428058fbdb83a9c26d99646c/src/.DS_Store

maxkfranz commented:
This should be named `popper`: https://github.com/hmtinc/cytoscape.js-popper/blob/b2be9e1c1b617885428058fbdb83a9c26d99646c/src/collection.js#L6

This should be named `popperRef`: https://github.com/hmtinc/cytoscape.js-popper/blob/b2be9e1c1b617885428058fbdb83a9c26d99646c/src/collection.js#L25

This should be named `popper`: https://github.com/hmtinc/cytoscape.js-popper/blob/b2be9e1c1b617885428058fbdb83a9c26d99646c/src/core.js#L6

This should be named `popperRef`: https://github.com/hmtinc/cytoscape.js-popper/blob/b2be9e1c1b617885428058fbdb83a9c26d99646c/src/core.js#L22

maxkfranz commented:
The top-level popper-object getting functions should be one line each.  There is unnecessary code duplication right now.

E.g. https://github.com/hmtinc/cytoscape.js-popper/blob/b2be9e1c1b617885428058fbdb83a9c26d99646c/src/core.js#L7-L17

They should basically just be `createPopperObject( popperRef( options ) )`.

maxkfranz commented:
This function should not exist:

https://github.com/hmtinc/cytoscape.js-popper/blob/b2be9e1c1b617885428058fbdb83a9c26d99646c/src/createBoundingBox.js#L9-L22

You should be storing a `getBoundingBox()` function or similar in the internal options object -- set in the top level `popperRef()` functions --- so you can just query that.

hmtinc commented:
@maxkfranz  
The following changes have been applied 
-`resolveUndefined()` has been replaced with assign()
- `getTargetInfo` has been simplified to reduce redundant code and has been moved to `core.js` and `collection.js`
- `Common.js` has been deleted
-  Last remaining .DS_Store has been removed
- High level functions in `core.js` and `collection.js` have been renamed
- `getPopperObjectDimensions()` has been replaced by a `dimensions()` function passed along in `userOptions`

maxkfranz commented on src/createBoundingBox.js:
```diff
@@ -0,0 +1,60 @@
+};
+
+//Wrap given bounding Box to match popper.js bounding box
+module.exports.getPopperBoundingBox = function (target, cy, getPosition, boundingBox) {
```
take into account

opts.getPosition()
opts.getDimensions()

maxkfranz replied:
opts.getRenderedPosition()

defaults (set top lvl) => model positions

maxkfranz commented on src/collection.js:
```diff
@@ -0,0 +1,57 @@
+
+  //Append a position function
+  if (!(userOptions.position)) {
+    userOptions.position = () => isNode ? target.renderedPosition() : target.midpoint();
```
midpoint needs to be in rendered co-ords

r = m * z + p

maxkfranz replied:
```js
math.modelToRenderedPosition = function( p, zoom, pan ){
  return {
    x: p.x * zoom + pan.x,
    y: p.y * zoom + pan.y
  };
};
```

hmtinc replied:
`midpoint()` is now converted into a rendered position

maxkfranz commented on src/createReferenceObject.js:
```diff
@@ -0,0 +1,45 @@
+
+    //Override if a reference override is provided
+    if (userOptions.refObject) {
+        refObject = userOptions.refObject;
```
rm

hmtinc replied:
Removed

hmtinc commented on src/collection.js:
```diff
@@ -0,0 +1,57 @@
+
+//Append element specific values to the options
+function appendValues(target, userOptions) {
+  userOptions = userOptions ? assign(userOptions) : {};
```
Removed line 29 and used assign to build a proper options object.

hmtinc commented on src/collection.js:
```diff
@@ -0,0 +1,57 @@
+  userOptions = userOptions ? assign(userOptions) : {};
+
+  //Append dimensions function
+  if (!(userOptions.dimensions)) {
```
Dimensions is now a default.

hmtinc commented on src/collection.js:
```diff
@@ -0,0 +1,57 @@
+  var iscyElement = !isCy;
+  var isNode = iscyElement && target.isNode();
+  var cy = isCy ? target : target.cy();
+
```
`getTargeInfo()` has now be removed

hmtinc commented on src/core.js:
```diff
@@ -0,0 +1,53 @@
+
+
+//Append element specific values to the options
+function appendValues(target, userOptions) {
```
Assign pattern is now used in appendValues

hmtinc commented on src/createBoundingBox.js:
```diff
@@ -0,0 +1,60 @@
+module.exports.getPopperBoundingBox = function (target, cy, getPosition, boundingBox) {
+    var position = getPosition(target);
+    var cyOffset = cy.container().getBoundingClientRect();
+
```
Variables set

hmtinc commented on src/createReferenceObject.js:
```diff
@@ -0,0 +1,45 @@
+//Create a reference object for an element
+module.exports.getRef = function (target, userOptions) {
+    //Determine element position
+    var position = userOptions.position;
```
Since `getRef()` just passes on `useroptions` now, there is no need to initialize variables anymore in `getRef()`

hmtinc commented:
@maxkfranz  I have applied the suggested changes

maxkfranz commented on src/core.js:
```diff
@@ -0,0 +1,41 @@
+      h: 3,
+    },
+    getDimensions : () => ({w: 3, h: 3}),
+    position : () => target.renderedPosition(),
```
`cy` doesn't have `cy.renderedPosition()`.  It probably just makes sense to return (0, 0) so the user would see something by default but know that he has to set `options.position` himself.

maxkfranz commented on src/createBoundingBox.js:
```diff
@@ -0,0 +1,65 @@
+    let scrollY = window.pageYOffset;
+    let scrollX = window.pageXOffset;
+   
+    //Exit if position is invalid
```
For what cases would the position be invalid?  Is it possible to set the defaults for `options` so that this doesn't happen by default?  If not, maybe we should just throw an `Error` since returning `undefined` would just cause an error farther away from the cause.

hmtinc replied:
Since position() can be a user defined function, its possible that the function could return an error as the user could technically pass a function that doesn't properly work with all targets. So, I thought It would be better to catch the error instead of letting something break further down. 

I have modified to the code to throw an error with a descriptive mesage

maxkfranz commented on src/createReferenceObject.js:
```diff
@@ -0,0 +1,27 @@
+        cyElement: target,
+
+        //Define the bounding box for the popper target
+        getBoundingClientRect: userOptions.boundingBox ? userOptions.boundingBox : function () {
```
Is there a case where we need `options.boundingBox`?  Or is it enough to use `options.position()` and `options.getDimensions()` to build the rect?

hmtinc replied:
`Options.boundingBox` allows the user to pass in a pre formatted bounding box into popper. The initial idea was that without the option to override the bounding box, if the user had a bounding box from elsewhere (e.g `.popperRef()), then they would have to manually extract `position()` and `getDimensions()`

Essentially it was added so you could just skip the process of having to define the `position()` function if you already had a bounding box from elsewhere.

maxkfranz commented on src/render.js:
```diff
@@ -0,0 +1,40 @@
+  }
+  catch (e) {
+    //Stop creating a popper
+    return;
```
Maybe we should just let exceptions propagate up

maxkfranz commented on src/collection.js:
```diff
@@ -31,6 +31,18 @@ function createOptionsObject(target, userOptions) {
+//Get the rendered center 
+function getRenderedCenter(target){
+  let pos = target.renderedPosition();
+  let offsetX = target.renderedWidth() / 2;
```
This should use `getDimensions()` so the user's definition of the element's dimensions is applied here.
